server/raft_transport.go

```diff
@@ -101,45 +112,106 @@ func (t *rpcTransport) Stop(id proto.RaftNodeID) {
+// via that client, exiting when the client fails or when it idles out. All
+// messages remaining in the queue at that point are lost and a new instance of
+// processQueue should be started by the next message to be sent.
+// TODO(tschottdorf) should let MultiRaft know if the node is down;
+// need a feedback mechanism for that. Potentially easiest is to arrange for
+// the next call to Send() to fail appropriately.
+func (t *rpcTransport) processQueue(raftNodeID proto.RaftNodeID) {
+	t.mu.Lock()
+	ch, ok := t.queues[raftNodeID]
+	t.mu.Unlock()
+	if !ok {
+		return
 	}
-
-	nodeID, _ := proto.DecodeRaftNodeID(proto.RaftNodeID(req.Message.To))
+	// Clean-up when the loop below shuts down.
+	defer func() {
+		t.mu.Lock()
+		delete(t.queues, raftNodeID)
+		t.mu.Unlock()
+	}()
+
+	nodeID, _ := proto.DecodeRaftNodeID(raftNodeID)
 	addr, err := t.gossip.GetNodeIDAddress(nodeID)
 	if err != nil {
-		return err
+		log.Errorf("could not get address for node %d: %s", nodeID, err)
 	}
-
 	client := rpc.NewClient(addr, nil, t.rpcContext)
 	select {
-	case <-client.Ready:
+	case <-t.rpcContext.Stopper.ShouldStop():
+		return
 	case <-client.Closed:
-		return util.Errorf("raft client failed to connect")
+		log.Warningf("raft client failed to connect")
+		return
+	case <-time.After(raftIdleTimeout):
+		// Should never happen.
+		log.Errorf("raft client stuck connecting")
+		return
+	case <-client.Ready:
 	}
 
-	call := client.Go(raftMessageName, protoReq, &proto.RaftMessageResponse{}, nil)
-	select {
-	case <-call.Done:
-		// If the call failed synchronously, report an error.
-		return call.Error
-	default:
-		// Otherwise, fire-and-forget.
-		go func() {
-			<-call.Done
+	done := make(chan *gorpc.Call, cap(ch))
+	var req *multiraft.RaftMessageRequest
+	protoReq := &proto.RaftMessageRequest{}
+	protoResp := &proto.RaftMessageResponse{}
+	for {
+		select {
+		case <-t.rpcContext.Stopper.ShouldStop():
+			return
+		case <-time.After(raftIdleTimeout):
+			return
+		case <-client.Closed:
+			log.Warningf("raft client closed")
+			return
+		case call := <-done:
 			if call.Error != nil {
 				log.Errorf("raft message failed: %s", call.Error)
 			}
-		}()
-		return nil
+			continue
+		case req = <-ch:
+		}
+		if req == nil {
+			return
+		}
+
+		// Convert to proto format.
+		protoReq.Reset()
+		protoReq.GroupID = req.GroupID
+		var err error
+		if protoReq.Msg, err = req.Message.Marshal(); err != nil {
+			log.Errorf("could not marshal message: %s", err)
+		}
+
+		if !client.IsHealthy() {
+			log.Warningf("client is unhealthy")
+			return
+		}
+		client.Go(raftMessageName, protoReq, protoResp, done)
 	}
 }
 
+// Send a message to the recipient specified in the request.
+func (t *rpcTransport) Send(req *multiraft.RaftMessageRequest) error {
+	raftNodeID := proto.RaftNodeID(req.Message.To)
+	t.mu.Lock()
+	ch, ok := t.queues[raftNodeID]
+	if !ok {
+		ch = make(chan *multiraft.RaftMessageRequest, raftSendBufferSize)
+		t.queues[raftNodeID] = ch
```

Comment:
we could avoid dropping messages at all by Unlock'ing at end of the function and not recreating the channel (we could put a tuple in the map indicating if it's being processed). I'm not sure that dropping messages while a node is down is bad though. Your call.